core/src/consistency.rs

```diff
@@ -0,0 +1,148 @@
+        }
+    }
+
+    pub fn new_pending(
+        event: ConsistencyEvent,
+        group: ConsistencyGroup,
+        pending_events: Vec<ConsistencyEvent>,
+    ) -> Self {
+        let pending = pending_events
+            .into_iter()
+            .map(|event| PendingConsistency {
+                event,
+                group: group.clone(),
+            })
+            .collect();
+        Self { event, pending }
+    }
+}
+
+#[derive(Clone, Debug, Serialize)]
+pub enum ConsistencyEvent {
+    // CAUSES
+    Publish(Address),                                   // -> Hold
+    AddPendingValidation(Address),                      // -> RemovePendingValidation
+    SignalZomeFunctionCall(snowflake::ProcessUniqueId), // -> ReturnZomeFunctionResult
+
+    // EFFECTS
+    Hold(Address),                                        // <- Publish
+    UpdateEntry(Address, Address),                        // <- Publish, entry_type=Update
+    RemoveEntry(Address, Address),                        // <- Publish, entry_type=Deletion
+    AddLink(Link),                                        // <- Publish, entry_type=LinkAdd
+    RemoveLink(Link),                                     // <- Publish, entry_type=LinkRemove
+    RemovePendingValidation(Address),                     // <- AddPendingValidation
+    ReturnZomeFunctionResult(snowflake::ProcessUniqueId), // <- SignalZomeFunctionCall
+}
+
+#[derive(Clone, Debug, Serialize)]
+struct PendingConsistency {
+    event: ConsistencyEvent,
+    group: ConsistencyGroup,
+}
+
+#[derive(Clone, Debug, Serialize)]
+pub enum ConsistencyGroup {
+    Source,
+    Validators,
+}
+
+impl ConsistencyModel {
+    pub fn process_action(&mut self, action: &Action) -> Option<ConsistencySignal> {
+        use ConsistencyEvent::*;
+        use ConsistencyGroup::*;
+        match action {
+            Action::Commit((entry, crud_link, _)) => {
+                let address = entry.address();
+                let hold = Hold(address.clone());
+                let meta = crud_link.clone().and_then(|crud| match entry {
+                    Entry::App(_, _) => Some(UpdateEntry(crud, address.clone())),
+                    Entry::Deletion(_) => Some(RemoveEntry(crud, address.clone())),
+                    Entry::LinkAdd(link_data) => Some(AddLink(link_data.clone().link)),
+                    Entry::LinkRemove(link_data) => Some(RemoveLink(link_data.clone().link)),
+                    // Question: Why does Entry::LinkAdd take LinkData instead of Link?
+                    _ => None,
+                });
+                let mut pending = vec![hold];
+                meta.map(|m| pending.push(m));
+                let signal =
+                    ConsistencySignal::new_pending(Publish(address.clone()), Validators, pending);
+                self.commit_cache.insert(address, signal);
+                None
+            }
+            Action::Publish(address) => self.commit_cache.remove(address).or_else(|| {
```

Comment:
Ok, I don't think I understand how this is meant to work. Above we construct a pending list including metas, but remove the whole `ConsistencySignal` including that list of meta-pendings when we see the according `Publish` ?

Reply:
@lucksus I should definitely add a comment. My assumption for dealing with private entries is thus: The signal that specifies that nodes should await a `Hold` is only fired on `Publish`, rather than on `Commit` as in nodejs_conductor. The complexity here is that the `Publish` action only contains an address, while most of the important data comes with the `Commit` action. 

So, when the consistency model sees a Commit, it constructs the `ConsistencySignal` that should be sent by a `Publish`, and stores it in this hashmap. The `remove` actually returns the value removed, it could really be called `fetch_and_remove`.

It occurs to me that private entries will clutter up the commit_cache since they'll never get published. So maybe I should check if they are publishable before even putting them in the cache.

Does that all sound right?

Reply:
Ah, that clarifies it! Sorry, got a bit confused. Yeah, private entries will fill the hash map, but I see how this works now.

Reply:
@lucksus just added a commit which checks for can_publish before caching signals, and makes this more readable.